bindings.c

```diff
@@ -4092,6 +4292,363 @@ static int proc_swaps_read(char *buf, size_t size, off_t offset,
+ * @cfd : the file descriptor of the mounted cgroup. eg: /sys/fs/cgroup/cpu
+ */
+static int calc_pid(char ***pid_buf, char *dpath, int depth, int sum, int cfd)
+{
+	DIR *dir;
+	int fd;
+	struct dirent *file;
+	FILE *f = NULL;
+	size_t linelen = 0;
+	char *line = NULL;
+	int pd;
+	char *path_dir, *path;
+	char **pid;
+
+	/* path = dpath + "/cgroup.procs" + /0 */
+	path = (char *)malloc(strlen(dpath) + 20);
+	if (!path) {
+		lxcfs_error("Failed to malloc %s in calc_pid!\n", dpath);
+		return sum;
+	}
+
+	strcpy(path, dpath);
+	fd = openat(cfd, path, O_RDONLY);
+	if (fd < 0)
+		goto out;
+
+	dir = fdopendir(fd);
+	if (dir == NULL)
+		goto out;
+
+	while (((file = readdir(dir)) != NULL) && depth > 0) {
+		if (strncmp(file->d_name, ".", 1) == 0)
+			continue;
+		if (file->d_type == DT_DIR) {
+			/* path + '/' + d_name +/0 */
+			path_dir = (char *)malloc(strlen(path) + 2 + sizeof(file->d_name));
+			if (!path_dir) {
+				lxcfs_error("Failed to malloc %s in calc_pid!\n", path);
+				continue;
+			}
+			strcpy(path_dir, path);
+			strcat(path_dir, "/");
+			strcat(path_dir, file->d_name);
+			pd = depth - 1;
+			sum = calc_pid(pid_buf, path_dir, pd, sum, cfd);
+			free(path_dir);
+		}
+	}
+	closedir(dir);
+
+	/* realloc() failed */
+	if (sum == -1)
+		goto out;
+
+	strcat(path, "/cgroup.procs");
+	fd = openat(cfd, path, O_RDONLY);
+	if (fd < 0)
+		goto out;
+
+	f = fdopen(fd, "r");
+	if (!f)
+		goto out;
+
+	while (getline(&line, &linelen, f) != -1) {
+		pid = (char **)realloc(*pid_buf, sizeof(char *) * (sum + 1));
+		if (pid == NULL) {
+			lxcfs_error("%s\n", "realloc error!");
+			for (; sum > 0; sum--)
+				free((*pid_buf)[sum - 1]);
+			sum = -1;
+			goto out;
+		}
+		*pid_buf = pid;
+		do {
+			*(*pid_buf + sum) = (char *)malloc(strlen(line) + 3);
+		} while (*(*pid_buf + sum) == NULL);
+		strcpy(*(*pid_buf + sum), line);
+		sum++;
+	}
+	fclose(f);
+out:
+	free(path);
+	return sum;
+}
+/*
+ * calc_load calculates the load according to the following formula:
+ * load1 = load0 * exp + active * (1 - exp)
+ *
+ * load1: the new loadavg.
+ * load0: the former loadavg.
+ * active: the total number of running pid at this moment.
+ * exp: the fixed-point defined in the beginning.
+ */
+static unsigned long
+calc_load(unsigned long load, unsigned long exp, unsigned long active)
+{
+	unsigned long newload;
+
+	active = active > 0 ? active * FIXED_1 : 0;
+	newload = load * exp + active * (FIXED_1 - exp);
+	if (active >= load)
+		newload += FIXED_1 - 1;
+
+	return newload / FIXED_1;
+}
+
+/*
+ * Return 0 means that container p->cg is closed.
+ * Return -1 means that error occurred in refresh.
+ * Positive num equals the total number of pid.
+ */
+static int refresh_load(struct load_node *p, char *path)
+{
+	FILE *f = NULL;
+	char **idbuf;
+	char proc_path[256];
+	int i, ret, run_pid = 0, total_pid = 0, last_pid = 0;
+	char *line = NULL;
+	size_t linelen = 0;
+	int sum, length;
+	DIR *dp;
+	struct dirent *file;
+
+	idbuf = (char **)malloc(sizeof(char *));
```

Comment:
Please don't cast the result of `malloc()`.

Reply:
OK.